File: src/11-challenge/solution.py
from typing import Final


def fn(warehouse: list[str]) -> int:
    gift: Final = "*"
    camera: Final = "#"
    unattended_gifts = 0

    for row_i, row in enumerate(warehouse):
        up_row = warehouse[row_i - 1] if (row_i - 1) > -1 else "." * len(row)
        down_row = (
            warehouse[row_i + 1] if (row_i + 1) < len(warehouse) else "." * len(row)
        )

        for col_i, col in enumerate(row):
            if col != gift:
                continue

            up_cell = up_row[col_i]
            down_cell = down_row[col_i]

            left_cell = row[col_i - 1] if (col_i - 1) > -1 else "."
            right_cell = row[col_i + 1] if (col_i + 1) < len(row) else "."

            cardinal_cells = [up_cell, down_cell, left_cell, right_cell]

            unattended_gifts += camera not in cardinal_cells

    return unattended_gifts

File: src/11-challenge/solution.ts
type Cell = '*' | '#' | '.'

export default function fn(warehouse: string[]): number {
	let unattendedGifts = 0

	for (let row = 0; row < warehouse.length; row++) {
		const upRow = warehouse[row - 1] ?? '.'.repeat(warehouse[row].length)
		const downRow = warehouse[row + 1] ?? '.'.repeat(warehouse[row].length)

		for (let col = 0; col < warehouse[row].length; col++) {
			const cell = warehouse[row][col] as Cell
			if (cell !== '*') continue

			const upCell = upRow[col]
			const downCell = downRow[col]

			const leftCell = warehouse[row][col - 1] ?? '.'
			const rightCell = warehouse[row][col + 1] ?? '.'

			const cardinalCells = [upCell, downCell, leftCell, rightCell]

			unattendedGifts += +!cardinalCells.includes('#')
		}
	}

	return unattendedGifts
}
